lib/pgauge_det_trace.cu

```diff
@@ -2,154 +2,81 @@
 #include <quda_matrix.h>
 #include <tune_quda.h>
 #include <gauge_field.h>
-#include <gauge_field_order.h>
-#include <launch_kernel.cuh>
 #include <comm_quda.h>
-#include <reduce_helper.h>
-#include <index_helper.cuh>
 #include <instantiate.h>
 
-namespace quda {
-
-  template <typename Float, int nColor_, QudaReconstructType recon_>
-  struct KernelArg : public ReduceArg<double2> {
-    static constexpr int nColor = nColor_;
-    static constexpr QudaReconstructType recon = recon_;
-    using real = typename mapper<Float>::type;
-    using Gauge = typename gauge_mapper<real, recon>::type;
-    int threads; // number of active threads required
-    int X[4]; // grid dimensions
-    int border[4];
-    Gauge dataOr;
-
-    KernelArg(const GaugeField &data) :
-      ReduceArg<double2>(),
-      dataOr(data),
-      threads(data.LocalVolumeCB())
-    {
-      for (int dir=0; dir<4; ++dir) {
-        border[dir] = data.R()[dir];
-        X[dir] = data.X()[dir] - border[dir]*2;
-      }
-    }
-  };
-
-  template <int blockSize, int type, typename Arg>
-  __global__ void compute(Arg arg)
-  {
-    int idx = threadIdx.x + blockIdx.x*blockDim.x;
-    int parity = threadIdx.y;
-
-    complex<double> val(0.0, 0.0);
-    while (idx < arg.threads) {
-      int X[4];
-#pragma unroll
-      for(int dr=0; dr<4; ++dr) X[dr] = arg.X[dr];
+#include <tunable_reduction.h>
+#include <kernels/gauge_det_trace.cuh>
 
-      int x[4];
-      getCoords(x, idx, X, parity);
-#pragma unroll
-      for(int dr=0; dr<4; ++dr) {
-        x[dr] += arg.border[dr];
-        X[dr] += 2*arg.border[dr];
-      }
-      idx = linkIndex(x,X);
-#pragma unroll
-      for (int mu = 0; mu < 4; mu++) {
-        Matrix<complex<typename Arg::real>, Arg::nColor> U = arg.dataOr(mu, idx, parity);
-        if (type == 0) val += getDeterminant(U);
-        else if (type == 1) val += getTrace(U);
-      }
-
-      idx += blockDim.x*gridDim.x;
-    }
-
-    arg.template reduce2d<blockSize,2>(val);
-  }
+namespace quda {
 
-  template <typename Float, int nColor, QudaReconstructType recon, int type>
-  class CalcFunc : TunableLocalParityReduction {
+  template <typename Float, int nColor, QudaReconstructType recon>
+  class CalcFunc : TunableReduction2D<> {
     double2 &result;
```

Comment:
`std::vector<double>` to match `gauge_plaq.cu`

Reply:
39b7227c9dd2c4372cee823121d124665c3baf6f